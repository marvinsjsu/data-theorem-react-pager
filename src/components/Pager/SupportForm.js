import React from 'react';
import PropTypes from 'prop-types';

export default class SupportForm extends React.Component {

  static propTypes = {
    sendMessage: PropTypes.func.isRequired,
    supportRequestUrl: PropTypes.string.isRequired
  };

  state = {
    name: '',
    email: '',
    message: ''
  };

  onSubmit = (e) => {
    e.preventDefault();
    console.log('onSubmit');
    const { sendMessage } = this.props;
    sendMessage({...this.state});
  };

  catchEnter = (e) => {

    if (e.key === 'Enter') {
      e.preventDefault();
    }
  };

  onChange = (key, e) => {
    this.setState({
      [key]: e.target.value
    });
  };

  render () {
    const { name, email, message } = this.state;
    const { closeSupportDialog } = this.props;

    return (
      <div className='container--form' id='support-form'>
        <form onSubmit={this.onSubmit}>
          <h4 className='title'>Support Request Form</h4>
          <button
            className="container--form__close"
            onClick={closeSupportDialog}
          >
            &times;
          </button>
          <div className='row'>
            <input
              id='name'
              className='input'
              type='text'
              name='name'
              value={name}
              placeholder='name'
              onChange={(e) => this.onChange('name', e)}
              onKeyDown={this.catchEnter}
              required
            />
          </div>

          <div className='row'>
            <input
              id='email'
              className='input'
              type='email'
              name='email'
              value={email}
              placeholder='email'
              onChange={(e) => this.onChange('email', e)}
              onKeyDown={this.catchEnter}
              required
            />
          </div>

          <div className='row'>
            <textarea
              id='message'
              className='textarea--message'
              name='message'
              value={message}
              placeholder='message'
              onChange={(e) => this.onChange('message', e)}
              onKeyDown={this.catchEnter}
              required
            />
          </div>

          <div className='row'>
            <button
              type='submit'
              className='btn'
              disabled={!name || !email || !message}
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }
};
