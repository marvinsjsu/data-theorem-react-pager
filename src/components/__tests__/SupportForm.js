import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SupportForm from '../Pager/SupportForm';

jest.mock('../../utils/api');
Enzyme.configure({ adapter: new Adapter() });

const supportRequestUrl = 'http://localhost:3000/api/v1/test';

describe(`The SupportForm component`, () => {

  it (`renders as expected`, () => {
    const component = renderer.create(
      <SupportForm
        supportRequestUrl={supportRequestUrl}
        closeSupportDialog={() => {}}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test (`user enters name in input and is shown`, () => {
    const wrapper = shallow(
      <SupportForm
        supportRequestUrl={supportRequestUrl}
        closeSupportDialog={() => {}}
      />
    );

    wrapper.find('#name').simulate('change', {
      target: { value: 'calvin' }
    });

    expect(wrapper.find('#name').props().value).toEqual('calvin');
  });

  test (`user enters email in input and is shown`, () => {
    const wrapper = shallow(
      <SupportForm
        supportRequestUrl={supportRequestUrl}
        closeSupportDialog={() => {}}
      />
    );

    wrapper.find('#email').simulate('change', {
      target: { value: 'calvin@pups.com' }
    });

    expect(wrapper.find('#email').props().value).toEqual('calvin@pups.com');
  });

  test (`user enters message in input and is shown`, () => {
    const wrapper = shallow(
      <SupportForm
        supportRequestUrl={supportRequestUrl}
        closeSupportDialog={() => {}}
      />
    );

    wrapper.find('#message').simulate('change', {
      target: { value: 'need a walk' }
    });

    expect(wrapper.find('#message').props().value).toEqual('need a walk');
  });

  test (`user submits support form, fires event preventsDefault`, () => {
    const wrapper = shallow(
      <SupportForm
        supportRequestUrl={supportRequestUrl}
        closeSupportDialog={() => {}}
      />
    );

    let prevented = false;
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {
        prevented = true;
      }
    });

    expect(prevented).toBe(true);
  });

  test (`user hits enter on input, fires event preventsDefault`, () => {
    let closed = false;
    const closeSupportDialog = () => { closed = true };
    const wrapper = shallow(
      <SupportForm
        supportRequestUrl={supportRequestUrl}
        closeSupportDialog={closeSupportDialog}
      />
    );

    let prevented = false;
    wrapper.find('#name').simulate('keydown', {
      key: 'Enter',
      preventDefault: () => {
        prevented = true;
      }
    });

    expect(prevented).toBe(true);
  });

  test (`user clicks on close (x) button`, () => {
    let closed = false;
    const closeSupportDialog = () => { closed = true };
    const wrapper = shallow(
      <SupportForm
        supportRequestUrl={supportRequestUrl}
        closeSupportDialog={closeSupportDialog}
      />
    );

    wrapper.find('.container--form__close').simulate('click');
    expect(closed).toBe(true);
  });

});