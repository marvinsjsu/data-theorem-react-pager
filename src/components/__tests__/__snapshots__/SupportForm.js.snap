// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SupportForm component renders as expected 1`] = `
<div
  className="container--form"
  id="support-form"
>
  <form
    onSubmit={[Function]}
  >
    <h4
      className="title"
    >
      Support Request Form
    </h4>
    <button
      className="container--form__close"
      onClick={[Function]}
    >
      Ã—
    </button>
    <div
      className="row"
    >
      <input
        className="input"
        id="name"
        name="name"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="name"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="input"
        id="email"
        name="email"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="email"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <textarea
        className="textarea--message"
        id="message"
        name="message"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="message"
        required={true}
        value=""
      />
    </div>
    <div
      className="row"
    >
      <button
        className="btn"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
